### Text Splitting from Documents-RecursiveCharacter Text Splitters

This text splitter is the recommended one for generic text.It is parameterized by a list of characters.It tries to split on them in order unitl the chunks are small enough.The defailt list is ["\n\n","\n"," ",""].This has the effect of trying to keep all the paragraphs(and then sentences,and then words) together as long as possible as these would generically seem to be the strongest semantically related pieces of text.

- How the text is split:List of Characters 
- How the chunk size is measured:by number of characters.


In [52]:
##Reading a pdf file 
from langchain_community.document_loaders import PyPDFLoader 
loader=PyPDFLoader('attention.pdf')
docs=loader.load() 
docs

[Document(metadata={'producer': 'pdfTeX-1.40.25', 'creator': 'LaTeX with hyperref', 'creationdate': '2024-04-10T21:11:43+00:00', 'author': '', 'keywords': '', 'moddate': '2024-04-10T21:11:43+00:00', 'ptex.fullbanner': 'This is pdfTeX, Version 3.141592653-2.6-1.40.25 (TeX Live 2023) kpathsea version 6.3.5', 'subject': '', 'title': '', 'trapped': '/False', 'source': 'attention.pdf', 'total_pages': 15, 'page': 0, 'page_label': '1'}, page_content='Provided proper attribution is provided, Google hereby grants permission to\nreproduce the tables and figures in this paper solely for use in journalistic or\nscholarly works.\nAttention Is All You Need\nAshish Vaswani∗\nGoogle Brain\navaswani@google.com\nNoam Shazeer∗\nGoogle Brain\nnoam@google.com\nNiki Parmar∗\nGoogle Research\nnikip@google.com\nJakob Uszkoreit∗\nGoogle Research\nusz@google.com\nLlion Jones∗\nGoogle Research\nllion@google.com\nAidan N. Gomez∗ †\nUniversity of Toronto\naidan@cs.toronto.edu\nŁukasz Kaiser∗\nGoogle Brain\nlukaszk

In [53]:
type(docs[0])

langchain_core.documents.base.Document

How to Recursively Split Text by Characters

In [54]:
from langchain_text_splitters import RecursiveCharacterTextSplitter 
 
text_splitter=RecursiveCharacterTextSplitter(chunk_size=500,chunk_overlap=50)
final_documents=text_splitter.split_documents(docs) 
final_documents

[Document(metadata={'producer': 'pdfTeX-1.40.25', 'creator': 'LaTeX with hyperref', 'creationdate': '2024-04-10T21:11:43+00:00', 'author': '', 'keywords': '', 'moddate': '2024-04-10T21:11:43+00:00', 'ptex.fullbanner': 'This is pdfTeX, Version 3.141592653-2.6-1.40.25 (TeX Live 2023) kpathsea version 6.3.5', 'subject': '', 'title': '', 'trapped': '/False', 'source': 'attention.pdf', 'total_pages': 15, 'page': 0, 'page_label': '1'}, page_content='Provided proper attribution is provided, Google hereby grants permission to\nreproduce the tables and figures in this paper solely for use in journalistic or\nscholarly works.\nAttention Is All You Need\nAshish Vaswani∗\nGoogle Brain\navaswani@google.com\nNoam Shazeer∗\nGoogle Brain\nnoam@google.com\nNiki Parmar∗\nGoogle Research\nnikip@google.com\nJakob Uszkoreit∗\nGoogle Research\nusz@google.com\nLlion Jones∗\nGoogle Research\nllion@google.com\nAidan N. Gomez∗ †\nUniversity of Toronto\naidan@cs.toronto.edu'),
 Document(metadata={'producer': 'pd

In [55]:
print(final_documents[0])
print(final_documents[1])

page_content='Provided proper attribution is provided, Google hereby grants permission to
reproduce the tables and figures in this paper solely for use in journalistic or
scholarly works.
Attention Is All You Need
Ashish Vaswani∗
Google Brain
avaswani@google.com
Noam Shazeer∗
Google Brain
noam@google.com
Niki Parmar∗
Google Research
nikip@google.com
Jakob Uszkoreit∗
Google Research
usz@google.com
Llion Jones∗
Google Research
llion@google.com
Aidan N. Gomez∗ †
University of Toronto
aidan@cs.toronto.edu' metadata={'producer': 'pdfTeX-1.40.25', 'creator': 'LaTeX with hyperref', 'creationdate': '2024-04-10T21:11:43+00:00', 'author': '', 'keywords': '', 'moddate': '2024-04-10T21:11:43+00:00', 'ptex.fullbanner': 'This is pdfTeX, Version 3.141592653-2.6-1.40.25 (TeX Live 2023) kpathsea version 6.3.5', 'subject': '', 'title': '', 'trapped': '/False', 'source': 'attention.pdf', 'total_pages': 15, 'page': 0, 'page_label': '1'}
page_content='University of Toronto
aidan@cs.toronto.edu
Łukasz Kaise

In [56]:
from langchain_community.document_loaders import TextLoader 

loader=TextLoader('speech.txt') 
docs=loader.load() 
print(docs)

[Document(metadata={'source': 'speech.txt'}, page_content='"Your time is limited, so don’t waste it living someone else’s life.\n\nDon’t be trapped by dogma — which is living with the results of other people’s thinking. Don’t let the noise of others’ opinions drown out your own inner voice. And most important, have the courage to follow your heart and intuition.\n\nThey somehow already know what you truly want to become. Everything else is secondary.\n\nWhen I was 17, I read a quote that went something like:\nIf you live each day as if it was your last, someday you’ll most certainly be right."')]


In [57]:
speech="" 
with open("speech.txt")as f:
    speech=f.read() 
speech 
text_splitter=RecursiveCharacterTextSplitter(chunk_size=50,chunk_overlap=10) 
text=text_splitter.create_documents([speech])
print(text[0]) 
print(text[1]) 



page_content='"Your time is limited, so don’t waste it living'
page_content='it living someone else’s life.'


In [58]:
type(text[0])

langchain_core.documents.base.Document

### Character Text Splitter 

This is the simplest method.This splits based on given character sequence,which defaults to "\n\n".Chunk length is measured by number of characters.

- How the text is split:by single character separator
- How the chunk size is measured:by number of characters



In [59]:
from langchain_community.document_loaders import TextLoader 

loader=TextLoader('speech.txt') 
docs=loader.load() 
docs

[Document(metadata={'source': 'speech.txt'}, page_content='"Your time is limited, so don’t waste it living someone else’s life.\n\nDon’t be trapped by dogma — which is living with the results of other people’s thinking. Don’t let the noise of others’ opinions drown out your own inner voice. And most important, have the courage to follow your heart and intuition.\n\nThey somehow already know what you truly want to become. Everything else is secondary.\n\nWhen I was 17, I read a quote that went something like:\nIf you live each day as if it was your last, someday you’ll most certainly be right."')]

In [60]:
from langchain_text_splitters import CharacterTextSplitter
text_splitter=CharacterTextSplitter(separator="\n\n",chunk_size=100,chunk_overlap=20)
text_splitter.split_documents(docs)


Created a chunk of size 233, which is longer than the specified 100


[Document(metadata={'source': 'speech.txt'}, page_content='"Your time is limited, so don’t waste it living someone else’s life.'),
 Document(metadata={'source': 'speech.txt'}, page_content='Don’t be trapped by dogma — which is living with the results of other people’s thinking. Don’t let the noise of others’ opinions drown out your own inner voice. And most important, have the courage to follow your heart and intuition.'),
 Document(metadata={'source': 'speech.txt'}, page_content='They somehow already know what you truly want to become. Everything else is secondary.'),
 Document(metadata={'source': 'speech.txt'}, page_content='When I was 17, I read a quote that went something like:\nIf you live each day as if it was your last, someday you’ll most certainly be right."')]

Created a chunk of size 233, which is longer than the specified 100 --> Because it is now able to find the seperator

In [61]:
speech="" 
with open("speech.txt") as f: 
    speech=f.read() 

text_splitter=CharacterTextSplitter(chunk_size=100,chunk_overlap=20) 
text=text_splitter.create_documents([speech]) 
print(text[0]) 
print(text[1])

Created a chunk of size 233, which is longer than the specified 100


page_content='"Your time is limited, so don’t waste it living someone else’s life.'
page_content='Don’t be trapped by dogma — which is living with the results of other people’s thinking. Don’t let the noise of others’ opinions drown out your own inner voice. And most important, have the courage to follow your heart and intuition.'


### How to Split by HTML header 

HTMLHeaderTextSPlitter is a structure-aware chunker that splits text at the HTML element level and adds metadata for each header "relavant" to any given chunk.It can return chunks element by element or combine elements with the same metadata,with the objectives of 
- Keeping related text grouped(more or less) semantically 
- Preserving context-rich information encoded in document structures.it can be used with other text splitters as part of a chunking pipeline

In [62]:
from langchain_text_splitters import HTMLHeaderTextSplitter 

html_string=html_content = """
<html>
  <head>
    <title>Learn HTMLTextSplitter</title>
  </head>
  <body>
    <h1>Welcome to HTMLTextSplitter</h1>
    <p>This tool helps you split HTML content intelligently for LLM-based applications.</p>

    <h2>Why Use HTMLTextSplitter?</h2>
    <p>Because plain text splitters don't preserve HTML structure, which can lead to context loss.</p>
    <p>HTMLTextSplitter keeps the logical flow intact.</p>

    <h2>Features</h2>
    <ul>
      <li>Preserves semantic blocks like headings and paragraphs</li>
      <li>Supports recursive chunking with overlap</li>
      <li>Works well with web scraping pipelines</li>
    </ul>

    <h2>Resources</h2>
    <p>Check out the <a href="https://docs.langchain.com">LangChain documentation</a> for more info.</p>

    <h2>Conclusion</h2>
    <p>Use this tool when dealing with rich HTML to improve retrieval and comprehension in LLM workflows.</p>
  </body>
</html>
"""


headers_to_split_on=[("h1","Header 1"), 
                     ("h2","Header 2"), 
                     ("h3","Header 3")]
html_splitter=HTMLHeaderTextSplitter(headers_to_split_on)
html_header_splits=html_splitter.split_text(html_string) 
html_header_splits

[Document(metadata={'Header 1': 'Welcome to HTMLTextSplitter'}, page_content='Welcome to HTMLTextSplitter'),
 Document(metadata={'Header 1': 'Welcome to HTMLTextSplitter'}, page_content='This tool helps you split HTML content intelligently for LLM-based applications.'),
 Document(metadata={'Header 1': 'Welcome to HTMLTextSplitter', 'Header 2': 'Why Use HTMLTextSplitter?'}, page_content='Why Use HTMLTextSplitter?'),
 Document(metadata={'Header 1': 'Welcome to HTMLTextSplitter', 'Header 2': 'Why Use HTMLTextSplitter?'}, page_content="Because plain text splitters don't preserve HTML structure, which can lead to context loss.  \nHTMLTextSplitter keeps the logical flow intact."),
 Document(metadata={'Header 1': 'Welcome to HTMLTextSplitter', 'Header 2': 'Features'}, page_content='Features'),
 Document(metadata={'Header 1': 'Welcome to HTMLTextSplitter', 'Header 2': 'Features'}, page_content='Preserves semantic blocks like headings and paragraphs  \nSupports recursive chunking with overlap  

In [63]:
url="https://plato.stanford.edu/entries/plato-ethics/" 

headers_to_split_on=[("h1","Header 1"), 
                     ("h2","Header 2"), 
                     ("h3","Header 3"), 
                     ("h4","Header 4"),] 

html_splitter=HTMLHeaderTextSplitter(headers_to_split_on)
html_header_splits=html_splitter.split_text_from_url(url) 
html_header_splits

[Document(metadata={}, page_content='End container NOTE: Script required for drop-down button to work (mirrors).  \nEnd header wrapper End content End footer  \nEnd header  \nEnd navigation End search  \nStanford Encyclopedia of Philosophy  \nMenu  \nBrowse  \nTable of Contents  \nWhat\'s New  \nRandom Entry  \nChronological  \nArchives  \nAbout  \nEditorial Information  \nAbout the SEP  \nEditorial Board  \nHow to Cite the SEP  \nSpecial Characters  \nAdvanced Tools  \nContact  \nSupport SEP  \nSupport the SEP  \nPDFs for SEP Friends  \nMake a Donation  \nSEPIA for Libraries  \nBegin article sidebar End article sidebar NOTE: Article content must have two wrapper divs: id="article" and id="article-content" End article NOTE: article banner is outside of the id="article" div. End article-banner  \nEntry Navigation  \nEntry Contents  \nBibliography  \nAcademic Tools  \nFriends PDF Preview  \nAuthor and Citation Info  \nBack to Top  \nEnd article-content  \nBEGIN ARTICLE HTML #aueditable D

### How to Split JSON Data

This JSON splitter splits json data while allowing control over chunk sizes.It traverses json data depth first and builds smaller json chunks.It attempts to keep nested json objects whole but will split them if needed to keep chunks between a min_chunk_size and the max_chunk_size. 

If the value is not a needed json,but rather a very large string the string will not be split.If you need a hard cap on the chunk size considering composing this with a Recursive Text Splitter on those chunks.There is an optional pre-processing step to split lists,by first converting them to json(dict) and then splitting them as such. 

- How the text is split:JSON value
- How the chunk size is measured:by number of characters

In [64]:
import json 
import requests 
 
json_data=requests.get("https://jsonplaceholder.typicode.com/users").json() 


In [65]:
json_data

[{'id': 1,
  'name': 'Leanne Graham',
  'username': 'Bret',
  'email': 'Sincere@april.biz',
  'address': {'street': 'Kulas Light',
   'suite': 'Apt. 556',
   'city': 'Gwenborough',
   'zipcode': '92998-3874',
   'geo': {'lat': '-37.3159', 'lng': '81.1496'}},
  'phone': '1-770-736-8031 x56442',
  'website': 'hildegard.org',
  'company': {'name': 'Romaguera-Crona',
   'catchPhrase': 'Multi-layered client-server neural-net',
   'bs': 'harness real-time e-markets'}},
 {'id': 2,
  'name': 'Ervin Howell',
  'username': 'Antonette',
  'email': 'Shanna@melissa.tv',
  'address': {'street': 'Victor Plains',
   'suite': 'Suite 879',
   'city': 'Wisokyburgh',
   'zipcode': '90566-7771',
   'geo': {'lat': '-43.9509', 'lng': '-34.4618'}},
  'phone': '010-692-6593 x09125',
  'website': 'anastasia.net',
  'company': {'name': 'Deckow-Crist',
   'catchPhrase': 'Proactive didactic contingency',
   'bs': 'synergize scalable supply-chains'}},
 {'id': 3,
  'name': 'Clementine Bauch',
  'username': 'Samantha

In [66]:
from langchain_text_splitters import RecursiveJsonSplitter

json_splitter=RecursiveJsonSplitter(max_chunk_size=300) 
json_chunks = json_splitter.split_json({"data": json_data})


In [67]:
json_chunks

[{'data': [{'id': 1,
    'name': 'Leanne Graham',
    'username': 'Bret',
    'email': 'Sincere@april.biz',
    'address': {'street': 'Kulas Light',
     'suite': 'Apt. 556',
     'city': 'Gwenborough',
     'zipcode': '92998-3874',
     'geo': {'lat': '-37.3159', 'lng': '81.1496'}},
    'phone': '1-770-736-8031 x56442',
    'website': 'hildegard.org',
    'company': {'name': 'Romaguera-Crona',
     'catchPhrase': 'Multi-layered client-server neural-net',
     'bs': 'harness real-time e-markets'}},
   {'id': 2,
    'name': 'Ervin Howell',
    'username': 'Antonette',
    'email': 'Shanna@melissa.tv',
    'address': {'street': 'Victor Plains',
     'suite': 'Suite 879',
     'city': 'Wisokyburgh',
     'zipcode': '90566-7771',
     'geo': {'lat': '-43.9509', 'lng': '-34.4618'}},
    'phone': '010-692-6593 x09125',
    'website': 'anastasia.net',
    'company': {'name': 'Deckow-Crist',
     'catchPhrase': 'Proactive didactic contingency',
     'bs': 'synergize scalable supply-chains'}},


In [68]:
for chunk in json_chunks[:3]: 
    print(chunk)

{'data': [{'id': 1, 'name': 'Leanne Graham', 'username': 'Bret', 'email': 'Sincere@april.biz', 'address': {'street': 'Kulas Light', 'suite': 'Apt. 556', 'city': 'Gwenborough', 'zipcode': '92998-3874', 'geo': {'lat': '-37.3159', 'lng': '81.1496'}}, 'phone': '1-770-736-8031 x56442', 'website': 'hildegard.org', 'company': {'name': 'Romaguera-Crona', 'catchPhrase': 'Multi-layered client-server neural-net', 'bs': 'harness real-time e-markets'}}, {'id': 2, 'name': 'Ervin Howell', 'username': 'Antonette', 'email': 'Shanna@melissa.tv', 'address': {'street': 'Victor Plains', 'suite': 'Suite 879', 'city': 'Wisokyburgh', 'zipcode': '90566-7771', 'geo': {'lat': '-43.9509', 'lng': '-34.4618'}}, 'phone': '010-692-6593 x09125', 'website': 'anastasia.net', 'company': {'name': 'Deckow-Crist', 'catchPhrase': 'Proactive didactic contingency', 'bs': 'synergize scalable supply-chains'}}, {'id': 3, 'name': 'Clementine Bauch', 'username': 'Samantha', 'email': 'Nathan@yesenia.net', 'address': {'street': 'Doug

In [69]:
## The splitter can also output documents

docs=json_splitter.create_documents(texts=[{"data": json_data}]) 
for doc in docs[:3]: 
    print(doc)


page_content='{"data": [{"id": 1, "name": "Leanne Graham", "username": "Bret", "email": "Sincere@april.biz", "address": {"street": "Kulas Light", "suite": "Apt. 556", "city": "Gwenborough", "zipcode": "92998-3874", "geo": {"lat": "-37.3159", "lng": "81.1496"}}, "phone": "1-770-736-8031 x56442", "website": "hildegard.org", "company": {"name": "Romaguera-Crona", "catchPhrase": "Multi-layered client-server neural-net", "bs": "harness real-time e-markets"}}, {"id": 2, "name": "Ervin Howell", "username": "Antonette", "email": "Shanna@melissa.tv", "address": {"street": "Victor Plains", "suite": "Suite 879", "city": "Wisokyburgh", "zipcode": "90566-7771", "geo": {"lat": "-43.9509", "lng": "-34.4618"}}, "phone": "010-692-6593 x09125", "website": "anastasia.net", "company": {"name": "Deckow-Crist", "catchPhrase": "Proactive didactic contingency", "bs": "synergize scalable supply-chains"}}, {"id": 3, "name": "Clementine Bauch", "username": "Samantha", "email": "Nathan@yesenia.net", "address": {"